askama_derive/src/parser/mod.rs

```diff
@@ -240,31 +240,38 @@ fn nested_parenthesis(i: &str) -> IResult<&str, ()> {
     }
 }
 
-fn path(i: &str) -> IResult<&str, Vec<&str>> {
-    let root = opt(value("", ws(tag("::"))));
-    let tail = separated_list1(ws(tag("::")), identifier);
+enum PathOrIdentifier<'a> {
+    Path(Vec<&'a str>),
+    Identifier(&'a str),
+}
 
-    match tuple((root, identifier, ws(tag("::")), tail))(i) {
-        Ok((i, (root, start, _, rest))) => {
-            let mut path = Vec::new();
-            path.extend(root);
+fn path_or_identifier(i: &str) -> IResult<&str, PathOrIdentifier<'_>> {
+    let root = ws(opt(tag("::")));
+    let tail = opt(many1(preceded(ws(tag("::")), identifier)));
+
+    let (i, (root, start, rest)) = tuple((root, identifier, tail))(i)?;
+    let rest = rest.as_deref().unwrap_or_default();
+
+    // The returned identifier can be assumed to be path if:
+    // - Contains both a lowercase and uppercase character, i.e. a type name like `None`
+    // - Doesn't contain any lowercase characters, i.e. it's a constant
+    // In short, if it contains any uppercase characters it's a path.
+    match (root, start, rest) {
+        (Some(_), start, tail) => {
+            let mut path = Vec::with_capacity(2 + tail.len());
+            path.push("");
             path.push(start);
             path.extend(rest);
-            Ok((i, path))
+            Ok((i, PathOrIdentifier::Path(path)))
         }
-        Err(err) => {
-            if let Ok((i, name)) = identifier(i) {
-                // The returned identifier can be assumed to be path if:
-                // - Contains both a lowercase and uppercase character, i.e. a type name like `None`
-                // - Doesn't contain any lowercase characters, i.e. it's a constant
-                // In short, if it contains any uppercase characters it's a path.
-                if name.contains(char::is_uppercase) {
-                    return Ok((i, vec![name]));
-                }
-            }
-
-            // If `identifier()` fails then just return the original error
-            Err(err)
+        (None, name, []) if !name.contains(char::is_uppercase) => {
```

Comment:
What tests break if we leave out this arm? Maybe move the comment to be directly before this arm, which seems to be the main thing that's affected?

Reply:
This arm tells variable names (no upper case characters) and struct names (at least one upper case character) apart. I guess every test will fail without it. :)